prophyc/generators/base.py

```diff
@@ -0,0 +1,119 @@
+import os
+from prophyc import model
+
+
+class GenerateError(Exception):
+    pass
+
+
+def _write_file(file_path, string):
+    with open(file_path, "w") as f:
+        f.write(string)
+
+
+def _make_path(output_dir, base_name, extension):
+    assert extension.startswith(".")
+    assert os.path.isdir(output_dir), "Output directory %s doesn't exist." % output_dir
+    return os.path.join(output_dir, base_name + extension)
+
+
+class GeneratorAbc(object):
+    """
+        Generator class implements one language that model is translated to.
+        Generator needs to map at least one translator to its file extension in top_level_translators.
+        And to implement check_nodes method.
+    """
+    top_level_translators = {}
+
+    def check_nodes(self, nodes):
+        """
+            Before a model is translated to given language it needs to be checked if it conforms to given
+            language requirements. It's a place to raise exceptions if needed.
+        """
+
+
+class GeneratorBase(GeneratorAbc):
+    def __init__(self, outpu_directory="."):
+        self.output_dir = outpu_directory
+
+    def serialize(self, nodes, base_name):
+        self.check_nodes(nodes)
+
+        for file_path, translator in self.prepare_translators(base_name):
+            file_content = translator(nodes, base_name)
+            _write_file(file_path, file_content)
+
+    def prepare_translators(self, base_name):
+        for extension, translator_class in self.top_level_translators.items():
+            file_path = _make_path(self.output_dir, base_name, extension)
+            yield file_path, translator_class()
+
+
+class BlockTranslatorBase(object):
```

Comment:
The same split between Abc and Base (and Base impl details) could be made here.
Abc class would serve as a place to specify requirements (and possibilities) for Derived classes.
```
TranslatorAbc
# provides requirements for concrete Translator implementation
    1. translation functions
        # may not implement full set
        translate_<node_classname>
    2. post processing
        # choose at most one, specified behavior when both provided (e.g. exception)
        block_template
        block_post_process
    3. child translators
        # may not exist at all
        prerequisite_translators

TranslatorBase
    # the only public method
    __call__

    # private functions, implementation details of TranslatorBase
    # consider prefixing all of them by underscore or extracting out of class
    # to make it clear that neither derived implementer nor end user need to worry about those
    translation_methods_map
    process_nodes
    prepend_newline
    _nodes_dispatcher
    _get_translation_handler
```